[MLIR][OpenMP] Lowering nontemporal clause to LLVM IR for SIMD directive

kaviya2510 opened the pull request:
This patch adds nontemporal metadata to loads and stores of nontemporal list items.

tblah commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -5265,10 +5265,62 @@ OpenMPIRBuilder::getOpenMPDefaultSimdAlign(const Triple &TargetTriple,
 }
 
+static void appendNontemporalVars(BasicBlock *Block,
+                                  SmallVector<Value *> &NontemporalVars) {
```
nit
```suggestion
                                  SmallVectorImpl<Value *> &NontemporalVars) {
```

This allows small vectors with different default allocation sizes

tblah commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -5265,10 +5265,62 @@ OpenMPIRBuilder::getOpenMPDefaultSimdAlign(const Triple &TargetTriple,
+/// Attach nontemporal metadata to the load/store instructions of nontemporal
+/// variables of \p Block
+static void addNonTemporalMetadata(BasicBlock *Block, MDNode *Nontemporal,
+                                   SmallVector<Value *> &NontemporalVars) {
```
nit
```suggestion
                                   SmallVectorImpl<Value *> &NontemporalVars) {
```

tblah commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -5265,10 +5265,62 @@ OpenMPIRBuilder::getOpenMPDefaultSimdAlign(const Triple &TargetTriple,
+}
+
+/// Attach nontemporal metadata to the load/store instructions of nontemporal
+/// variables of \p Block
```
Please could you expand this comment explaining the logic in this function in a way that somebody not familiar with flang runtime descriptors could understand (as this is shared code).

tblah commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -5265,10 +5265,62 @@ OpenMPIRBuilder::getOpenMPDefaultSimdAlign(const Triple &TargetTriple,
                                 Value *IfCond, OrderKind Order,
-                                ConstantInt *Simdlen, ConstantInt *Safelen) {
+                                ConstantInt *Simdlen, ConstantInt *Safelen,
+                                SmallVector<Value *> NontemporalVars) {
```
If you want to avoid a copy:
```suggestion
                                SmallVectorImpl<Value *> &NontemporalVars) {
```

If you want to keep the copy (which I think is okay in this case because it does feel a bit weird if OpenMPIRBuilder modifies this argument)
```suggestion
                               ArrayRef<Value *> NontemporalVarsIn) {
  SmallVector<Value *> NontemporalVars{NontemporalVarsIn};
```

More info here https://llvm.org/docs/ProgrammersManual.html#llvm-adt-arrayref-h

tblah reviewed (commented):
Thanks for updating the PR

kaviya2510 committed:
[MLIR][OpenMP] Lowering nontemporal clause to LLVM IR for SIMD directive

kaviya2510 commented:
Thanks for the review @tblah. 
I addressed all the comments. Kindly let me know if it needs any mofications.

kiranchandramohan commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -5279,10 +5279,86 @@ OpenMPIRBuilder::getOpenMPDefaultSimdAlign(const Triple &TargetTriple,
+ * address of each individual array elements and perform load or store
+ * operation on it. Mark those load and store instructions as non-temporal.
+ *
+ * Case 3: Allocatable array
+ *     For an allocatable array, which might involve runtime type descriptor,
+ * needs to navigate through descriptors using two or more GEP and load
+ * instructions to compute the address of each individual element in an array.
+ * Mark those load or store which access the individual array elements as
```
Allocatable array is a Fortran concept and ideally we do not want OpenMPIRBuilder to be aware of these. Is it possible to avoid special handling for Fortran descriptors?

kaviya2510 replied:
Hi @kiranchandramohan,
Thanks for the comments.
If we ignore the case for allocatable arrays(Fortran descriptors), testcases using such allocatable arrays will not include the relevant metadata(nontemporal) in the generated IR. In clang, for dynamically allocated arrays they are adding the metadata as nontemporal  by directly access the alloca pointer of dynamic arrays for load and store instructions.

jsjodin replied:
Can we create a callback for attaching metadata? Callbacks are usually how front end specific things are handled, unless it is something more generic and involved that the OpenMPIRBuilder has to keep track of.

kaviya2510 replied:
Hi @jsjodin,

Apologies for the delayed response. Yes, I think using a callback is helpful in this scenario. I have implemented the translation on the OpenMPToLLVMIRTranslation file and replaced the nontemporal function with callbacks in OMPIRBuilder.

kaviya2510 commented:
@kiranchandramohan , ping for review!

kiranchandramohan commented:
> @kiranchandramohan , ping for review!

Apologies for the delay from my side. I was mostly thinking about the following two points.

1. We should not be making aware the OpenMPIRBuilder code about Fortran specific stuff (like allocatables and descriptors) since this is neutral code and code shared between Clang and MLIR/Flang. Also, it is not clear whether it is possible to accurately identify the descriptor and array loads.
2. The conversion from FIR to LLVM dialect knows which loads are for the descriptors and which are for the actual array contents. Since the LLVM Dialect in MLIR models the nontemporal attribute (https://github.com/llvm/llvm-project/blob/ad38e24eb74e97148faec97c4f843b87768b6e9b/mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td#L371), we could place the non-temporal attribute during this conversion. But this again has two issues because we are making FIR to LLVM aware of OpenMP semantics and also skipping the OpenMP IRBuilder for lowering OpenMP code.

CC: @kparzysz @skatrak @tblah @agozillon

kaviya2510 commented:
No Problem. Thanks for sharing your insights regarding handling of allocatable arrays.
I will explore the suggested solutions and get back.

kaviya2510 committed:
[MLIR][OpenMP] Created callback function for adding the metadata of nontemporal and handled the translation in OpenMPToLLVMIRTranslation.cpp

kaviya2510 commented:
Ping for review!

kiranchandramohan commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -2527,6 +2541,72 @@ convertOmpSimd(Operation &opInst, llvm::IRBuilderBase &builder,
+      if (llvm::LoadInst *li = dyn_cast<llvm::LoadInst>(&I)) {
+        if (!(li->getType()->isPointerTy()))
+          mem_ptr = li->getPointerOperand();
+      } else if (llvm::StoreInst *si = dyn_cast<llvm::StoreInst>(&I))
+        mem_ptr = si->getPointerOperand();
```
Nit: Add braces here.

kiranchandramohan commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -2527,6 +2541,72 @@ convertOmpSimd(Operation &opInst, llvm::IRBuilderBase &builder,
+   * Nontemporal variables may be a scalar, fixed size or allocatable
+   * or pointer array
+   *
+   * Example scenarios for nontemporal variables:
+   * Case 1: Scalar variable
+   *    If the nontemporal variable is a scalar, it is allocated on stack.Load
+   * and store instructions directly access the alloca pointer of the scalar
+   * variable for fetching information about scalar variable or  writing
+   * into the scalar variable. Mark those load and store instructions as
+   * non-temporal.
+   *
+   * Case 2: Fixed Size array
+   *    If the nontemporal variable is a fixed-size array, it is allocated
+   * as a contiguous block of memory. It uses one GEP instruction, to compute
+   * the address of each individual array elements and perform load or store
+   * operation on it. Mark those load and store instructions as non-temporal.
+   *
+   * Case 3: Allocatable array
+   *     For an allocatable array, which might involve runtime type descriptor,
+   * needs to navigate through descriptors using two or more GEP and load
+   * instructions to compute the address of each individual element in an array.
+   * Mark those load or store which access the individual array elements as
+   * non-temporal.
+   */
```
Are these comments still applicable?

kaviya2510 replied:
Thanks for the comments.
Yes, all the three cases are still applicable. I moved the entire implementation from OpenMPIRBuilder to OpenMPToLLVMIRTranslation.cpp and made this changes specific to flang.

kaviya2510 commented:
I have implemented a callback function in Flang to add metadata for the nontemporal clause and handled the translation . Currently, I am passing an empty function as the default argument for the callback. Clang utilizes this for now, and it will be replaced by Clang's implementation in the future.

kaviya2510 committed:
Merge branch 'main' into nt_to_llvmir

tblah commented:
I'm not sure that a callback to MLIR is fully addressing Kiran's concern, because the MLIR dialect is also shared with things that are not just Fortran. This is a difficult problem to solve due to the layering of different parts of the compiler and I can't see a nice way to do this.

@kiranchandramohan which do you think is the least bad option of the two you suggested previously?
1. Continuing with special handling of boxes (and the regular memcpys between the boxes) in OpenMPToLLVMIRTranslation
2. Your suggestion to do this in the conversion from FIR to LLVM dialect?

kiranchandramohan commented:
> I'm not sure that a callback to MLIR is fully addressing Kiran's concern, because the MLIR dialect is also shared with things that are not just Fortran. This is a difficult problem to solve due to the layering of different parts of the compiler and I can't see a nice way to do this.
> 
> @kiranchandramohan which do you think is the least bad option of the two you suggested previously?
> 
> 1. Continuing with special handling of boxes (and the regular memcpys between the boxes) in OpenMPToLLVMIRTranslation
> 2. Your suggestion to do this in the conversion from FIR to LLVM dialect

Thanks @tblah for chiming in here.

Thinking about this again, the loads and stores on the real data are always visible in HLFIR/FIR code. I assume the current problem is that the MLIR value captured in the temporal clause is for the descriptor (or its address) and not for the address of the real data. If we have a pass in Flang that moves the temporal clause values from the descriptor to the real address of the data, then we need not do reverse engineering in the translation to find the correct loads and stores.

For the following example, it will be moving from `nontemporal(%3)` to `nontemporal(%18)`

```
    omp.simd nontemporal(%3 : !fir.ref<!fir.box<!fir.heap<!fir.array<?xi32>>>>) private(@_QFsbEi_private_i32 %2 -> %arg2 : !fir.ref<i32>) {
      omp.loop_nest (%arg3) : i32 = (%c1_i32) to (%c100_i32) inclusive step (%c1_i32) {
        %11 = fir.declare %arg2 {uniq_name = "_QFsbEi"} : (!fir.ref<i32>) -> !fir.ref<i32>
        fir.store %arg3 to %11 : !fir.ref<i32>
        %12 = fir.load %3 : !fir.ref<!fir.box<!fir.heap<!fir.array<?xi32>>>>
        %13 = fir.load %11 : !fir.ref<i32>
        %14 = fir.convert %13 : (i32) -> i64
        %15 = fir.box_addr %12 : (!fir.box<!fir.heap<!fir.array<?xi32>>>) -> !fir.heap<!fir.array<?xi32>>
        %16:3 = fir.box_dims %12, %c0 : (!fir.box<!fir.heap<!fir.array<?xi32>>>, index) -> (index, index, index)
        %17 = fir.shape_shift %16#0, %16#1 : (index, index) -> !fir.shapeshift<1>
        %18 = fir.array_coor %15(%17) %14 : (!fir.heap<!fir.array<?xi32>>, !fir.shapeshift<1>, i64) -> !fir.ref<i32>
        %19 = fir.load %18 : !fir.ref<i32>
        %20 = fir.load %4 : !fir.ref<i32>
        %21 = arith.addi %19, %20 : i32
        fir.store %21 to %18 : !fir.ref<i32>
        omp.yield
      }
    }
```    

Original Fortran.
```
subroutine sb(x, y)
  integer, allocatable :: x(:)
  integer :: y
  allocate(x(100))
  !$omp simd nontemporal(x)
  do i=1,100
  x(i) = x(i) + y
  end do
  !$omp end simd
end subroutine
```

Does such a pass make sense?

tblah commented:
Thanks for taking a look Kiran.

I don't think we can do nontemporal(%18) because %18 is not defined at the point it is being reffered to here. But we could have a flang pass which adds a non-temporal attribute to fir.load and fir.store, and that attribute could be propagated to the llvm operation.

kiranchandramohan commented:
> Thanks for taking a look Kiran.
> 
> I don't think we can do nontemporal(%18) because %18 is not defined at the point it is being refereed to. But we could have a flang pass which adds a non-temporal attribute to fir.load and fir.store, and that attribute could be propagated to the llvm operation.
> 
> For example,
> 
> ```
> 
> 
> For the following example, it will be moving from nontemporal(%3) to nontemporal(%18)
> 
>     omp.simd nontemporal(%3 : !fir.ref<!fir.box<!fir.heap<!fir.array<?xi32>>>>) private(@_QFsbEi_private_i32 %2 -> %arg2 : !fir.ref<i32>) {
>       omp.loop_nest (%arg3) : i32 = (%c1_i32) to (%c100_i32) inclusive step (%c1_i32) {
>         %11 = fir.declare %arg2 {uniq_name = "_QFsbEi"} : (!fir.ref<i32>) -> !fir.ref<i32>
>         fir.store %arg3 to %11 : !fir.ref<i32>
>         %12 = fir.load %3 : !fir.ref<!fir.box<!fir.heap<!fir.array<?xi32>>>>
>         %13 = fir.load %11 : !fir.ref<i32>
>         %14 = fir.convert %13 : (i32) -> i64
>         %15 = fir.box_addr %12 : (!fir.box<!fir.heap<!fir.array<?xi32>>>) -> !fir.heap<!fir.array<?xi32>>
>         %16:3 = fir.box_dims %12, %c0 : (!fir.box<!fir.heap<!fir.array<?xi32>>>, index) -> (index, index, index)
>         %17 = fir.shape_shift %16#0, %16#1 : (index, index) -> !fir.shapeshift<1>
>         %18 = fir.array_coor %15(%17) %14 : (!fir.heap<!fir.array<?xi32>>, !fir.shapeshift<1>, i64) -> !fir.ref<i32>
>         %19 = fir.load %18 : !fir.ref<i32> nontemporal
>         %20 = fir.load %4 : !fir.ref<i32>
>         %21 = arith.addi %19, %20 : i32
>         fir.store %21 to %18 : !fir.ref<i32> nontemporal
>         omp.yield
>       }
> ```

Yes, transferring to the correct loads and stores in a Flang pass and then handling the load/store with the attributes in translation/OpenMP seems to be a reasonable approach. 

There are a few ways of doing this:
1. We could consider adding an OpenMP dialect attribute for nontemporal loads and stores and then translate it to the nontemporal metadata during OpenMP translation
     -> at convertOperation of simd
     -> using the amendOperation interface.
2. We can add a nontemporal attribute to FIR loads/store and since LLVM dialect loads and stores have `UnitAttr:$nontemporal`, we could set this during the FIR to LLVM conversion.
3. Add an LLVM dialect attribute for nontemporal and then use that in fir and llvm loads and store.

The issue with attributes is that there is a danger that they can be dropped. But I assume we can guarantee retaining them in Flang.  If losing attributes is a concern, we could also consider creating a wrapper operation for the fir.load or fir.store. 
```
%t = omp.nontemporal {
  %1 = fir.load
  omp.yield %1
}
```

kaviya2510 commented:
Hi @kiranchandramohan , @tblah,
    Thanks for the suggestion. Could you please share a reference for adding a pass in LLVM as well as for adding an attribute to OpenMP? That would be helpful for me to analyze and implement it.

tblah commented:
> Thanks for the suggestion. Could you please share a reference for adding a pass in LLVM as well as for adding an attribute to OpenMP? That would be helpful for me to analyze and implement it.

I believe Kiran meant a pass in flang rather than in llvm. Here is an example of a flang pass for OpenMP: https://github.com/llvm/llvm-project/blob/main/flang/lib/Optimizer/OpenMP/MarkDeclareTarget.cpp, there are some more changes needed outside the new file so it would be a good idea to look at the commit which added the pass: https://github.com/llvm/llvm-project/commit/062fce6f4dd142cf9275abb3358f88b61a2db4e8

Here is a flang pass which adds attributes to some functions: https://github.com/llvm/llvm-project/blob/main/flang/lib/Optimizer/Transforms/FunctionAttr.cpp. Your pass would be similar except you would be adding the nontemporal attribute to some loads and stores.

For more information about MLIR passes, see the MLIR documentation: https://mlir.llvm.org/docs/PassManagement/

Please feel free to message me on the Flang slack if you need any support.

kaviya2510 commented:
Thanks for the references @tblah.

kaviya2510 committed:
Merge branch 'main' into nt_to_llvmir

kaviya2510 committed:
Lowering Nontemporal clause to LLVM IR: Added a Flang pass lower-nontemporal, which adds nontemporal attribute to the load and store of nontemporal variables during fir conversion

kaviya2510 commented:
Hi @kiranchandramohan and @tblah,
I added a flang pass "lower-nontemporal" which adds nontemporal attribute to FIR loads and stores. Kindly take a look at the recent patch and let me know if it needs any modifications. Thankyou!

tblah commented on flang/include/flang/Optimizer/OpenMP/Passes.td:
```diff
@@ -81,6 +81,13 @@ def DoConcurrentConversionPass : Pass<"omp-do-concurrent-conversion", "mlir::fun
   ];
 }
 
+def LowerNontemporalPass : Pass<"lower-nontemporal", "mlir::func::FuncOp"> {
```
Why not run this on `omp::SimdOp`? Running it on functions is wasteful because many functions will not contain a simd op and runs the risk of missing any simd ops which are not nested inside of functions (I can't see how that would happen currently but we have had problems before lowering non-function (but function-like) operations such as `omp.declare_reduction` because subsequent passes incorrectly assumed all operations were nested inside of functions.

tblah replied:
It was only a hypothetical example - I can't think of a case where we would do simd not inside of a function either.

Still, changing this to
```
def LowerNontemporalPass : Pass<"lower-nontemporal", "mlir::omp::SimdOp"> {
```
should be a better fit with the intention of the pass, because it will not modify anything outside of the simd op

tblah replied:
Re: declaration operations: It could be but I thought that would be so unlikely that it would be better to only make that (minimal) change if it was ever needed.

Re: this thread: Kaviya and I were unable to get it working. It turns out that the MLIR pass manager doesn't allow for arbitrary nesting: only 1 level deep. So you can run a func.func pass on a pass manager for a module because the func.func is a direct descendant. But it couldn't be done for simd because that could be anywhere in the operation tree. Sorry, I should have posted something here.

tblah commented on flang/lib/Optimizer/Passes/Pipelines.cpp:
```diff
@@ -274,8 +274,10 @@ void createHLFIRToFIRPassPipeline(mlir::PassManager &pm, bool enableOpenMP,
-  if (enableOpenMP)
+  if (enableOpenMP) {
     pm.addPass(flangomp::createLowerWorkshare());
+    pm.addPass(flangomp::createLowerNontemporalPass());
```
I think it would be better to include this pass much later in the pipeline so that loads and stores created by later passes are also included. Maybe around when the function attribute pass runs (right before FIR is lowered to LLVM).

There are also loads/stores created only during CodeGen (FIR to LLVM), but these are for boxes and so far as I can tell nontemporal only refers to data.

tblah reviewed (commented):
Thank you for rewriting your patch. This looks much better so far, but I have some more minor comments.

kaviya2510 commented on flang/lib/Optimizer/OpenMP/LowerNontemporal.cpp:
```diff
@@ -0,0 +1,71 @@
+          Operand = getBaseOperand(Operand);
+          if (is_contained(nontemporalOrigVars, Operand)) {
+            // Set the attribute
+            op->setAttr("nontemporal", UnitAttr::get(op->getContext()));
```
I have a small doubt here.
You suggested adding a new attribute "nontemporal" to the LLVM dialect and using it in the pass to mark loads and stores accordingly. 
However, LLVM dialect already supports the nontemporal attribute on load and store operations. Can't we use the existing attribute instead of introducing a new one? 

Could you please help me understand this better?

tblah replied:
Yes if an attribute already exists in the LLVM dialect it would be better to re-use it. Thanks for checking.

kaviya2510 replied:
Thankyou. I had one follow-up doubt I am hoping to get clarity on.
The nontemporal attribute in llvm dialect is defined specifically on LLVM_LoadOp and LLVM_StoreOp . Is it possible to reuse this same attribute directly in the FIR dialect (fir.load and fir.store) to mark operations as nontemporal?

My current idea is to define a nontemporal attribute in the FIR dialect(fir_LoadOp, fir_StoreOp) and use it to set the attribute at fir level. Would this be the right approach?

If not, could you please share your thoughts and provide further clarification?

tblah replied:
Okay yeah I guess as this is just a UnitAttr it isn't worth defining it carefully in the dialect.

So yes adding to fir.load and fir.store sounds okay to me. Please do that in a separate PR so it is to get review from other maintainers of FIR.

kaviya2510 replied:
Sure, I will do it. Thanks for the reply.

kaviya2510 committed:
Merge branch 'main' into nt_to_llvmir

kaviya2510 committed:
Addressed review comments : Added nontemporal attribute to fir.load and fir.store and used that to mark the operations as nontemporal

kaviya2510 commented on flang/lib/Optimizer/OpenMP/LowerNontemporal.cpp:
```diff
@@ -0,0 +1,71 @@
+    : public flangomp::impl::LowerNontemporalPassBase<LowerNontemporalPass> {
+  void addNonTemporalAttr(omp::SimdOp simdOp) {
+    if (!simdOp.getNontemporalVars().empty()) {
+      llvm::SmallVector<mlir::Value> nontemporalOrigVars;
```
Thanks for pointing that out. I have updated the latest patch to use simdOp.getNontemporalVars() directly in the is_contained check and the copy nontemporalOrigVars is not required.

kaviya2510 commented:
Hi @tblah , I have addressed all your review comments. The latest patch includes the following changes:
1. Added nontemporal attribute to fir.load and fir.store op in FIR dialect.
2. Marked the corresponding load and store operations as nontemporal using the setNontemporal() utility function at the FIR level.
3. Applied the "lower-nontemporal" pass just before FIRToLLVM conversion pass.

tblah commented on flang/lib/Optimizer/OpenMP/LowerNontemporal.cpp:
```diff
@@ -0,0 +1,76 @@
+        return operand;
+
+      Operation *definingOp = operand.getDefiningOp();
+      if (definingOp) {
+        for (Value srcOp : definingOp->getOperands()) {
+          return getBaseOperand(srcOp);
+        }
+      }
```
Which operation(s) are you intending to scan through here? I'm worried that the first operand on the defining op might not be the right one to return for all possible operations.

kaviya2510 replied:
```
  omp.simd nontemporal(%3 : !fir.ref<!fir.box<!fir.heap<!fir.array<?xi32>>>>)private(@_QFsimd_nontemporal_allocatableEi_private_i32 %2 -> %arg2 : !fir.ref<i32>) {
      omp.loop_nest (%arg3) : i32 = (%c1_i32) to (%c100_i32) inclusive step (%c1_i32) {
        %16 = fir.declare %arg2 {uniq_name = "_QFsimd_nontemporal_allocatableEi"} : (!fir.ref<i32>) -> !fir.ref<i32>
        fir.store %arg3 to %16 : !fir.ref<i32>
        %17 = fir.load %3 : !fir.ref<!fir.box<!fir.heap<!fir.array<?xi32>>>>
        %18 = fir.load %16 : !fir.ref<i32>
        %19 = fir.convert %18 : (i32) -> i64
        %20 = fir.box_addr %17 : (!fir.box<!fir.heap<!fir.array<?xi32>>>) -> !fir.heap<!fir.array<?xi32>>
        %c0_0 = arith.constant 0 : index
        %21:3 = fir.box_dims %17, %c0_0 : (!fir.box<!fir.heap<!fir.array<?xi32>>>, index) -> (index, index, index)
        %22 = fir.shape_shift %21#0, %21#1 : (index, index) -> !fir.shapeshift<1>
        %23 = fir.array_coor %20(%22) %19 : (!fir.heap<!fir.array<?xi32>>, !fir.shapeshift<1>, i64) -> !fir.ref<i32>
        %24 = fir.load %23 : !fir.ref<i32>
        %25 = fir.load %4 : !fir.ref<i32>
        %26 = arith.addi %24, %25 : i32
        %27 = fir.load %3 : !fir.ref<!fir.box<!fir.heap<!fir.array<?xi32>>>>
        %28 = fir.load %16 : !fir.ref<i32>
        %29 = fir.convert %28 : (i32) -> i64
        %30 = fir.box_addr %27 : (!fir.box<!fir.heap<!fir.array<?xi32>>>) -> !fir.heap<!fir.array<?xi32>>
        %c0_1 = arith.constant 0 : index
        %31:3 = fir.box_dims %27, %c0_1 : (!fir.box<!fir.heap<!fir.array<?xi32>>>, index) -> (index, index, index)
        %32 = fir.shape_shift %31#0, %31#1 : (index, index) -> !fir.shapeshift<1>
        %33 = fir.array_coor %30(%32) %29 : (!fir.heap<!fir.array<?xi32>>, !fir.shapeshift<1>, i64) -> !fir.ref<i32>
        fir.store %26 to %33 : !fir.ref<i32>
        omp.yield
      }
    }

```
My goal is to trace back and identify the base operand for fir.load or fir.store operation. 
Above mlir, which is an example of an allocatable array,  both `fir.load %24 = fir.load %23 : !fir.ref<i32>` and `fir.store fir.store %26 to %33 : !fir.ref<i32>`  operate on a single memory reference (the first operand), so I focused on recursively analyzing only the first operand of their defining operations to resolve the base.

tblah replied:
So you need to support fir.array_coor, fir.box_addr, and maybe some others (fir.convert...)? I can see this usage is correct for those operations but maybe it would be safer to limit to only working on these particular operations.

I am imagining something like this (although not all of these operations will be appropriate here): https://github.com/llvm/llvm-project/blob/main/flang/lib/Optimizer/Analysis/AliasAnalysis.cpp#L530

kaviya2510 replied:
I agree, but there are some issues with this approach.
I followed the above suggested method in my previous patch, where I added the `LowerNontemporalPass` after the `LowerWorkshare` pass. This pass did not work as expected when I moved it just above `FIRToLLVM` conversion pass. This is because the defining operations for the source operand of a `fir.load` instruction varies depending on where the pass is inserted in the pipeline.

In the first scenario, the source operand for the load instruction is from `fir.array_coor` whereas in the second scenario it comes from `fircg.ext_array_coor`. Inorder to resolve the issue, I followed this generic way which recursively walks back through operands to determine the base instruction(`fir.alloca`, `fir.declare`). This ensures that future changes in the mlir won’t break the pass.

What is your thoughts on this? 
Is it fine to proceed with the generalized approach or is it be better to explicitly handle each operation type?

kaviya2510 replied:
Thanks for the clarification. I will rewrite the patch as suggested and get back soon.

tblah commented on flang/lib/Optimizer/OpenMP/LowerNontemporal.cpp:
```diff
@@ -0,0 +1,76 @@
+        operand = getBaseOperand(operand);
+
+        if (llvm::is_contained(simdOp.getNontemporalVars(), operand)) {
+          if (auto loadOp = llvm::dyn_cast<fir::LoadOp>(op))
+            loadOp.setNontemporal(true);
+          else if (auto storeOp = llvm::dyn_cast<fir::StoreOp>(op))
+            storeOp.setNontemporal(true);
```
It would be more efficient to compute `operand` on if `op` is a load or a store operation. Doing it this way around means that `operand` is calculated every time, even if it will not be needed.

tblah commented on flang/test/Lower/OpenMP/simd-nontemporal.mlir:
```diff
@@ -0,0 +1,101 @@
+        fir.store %arg2 to %11 : !fir.ref<i32>
+        // CHECK:  %[[LOAD:.*]] = fir.load %[[A_DECL]] {nontemporal} : !fir.ref<i32>
+        %12 = fir.load %2 : !fir.ref<i32>
+        %13 = fir.load %4 : !fir.ref<i32>
```
Please could you add a CHECK line to show this one doesn't have the attribute added

kaviya2510 replied:
Sure, I will add the check.

tblah commented on flang/lib/Optimizer/CodeGen/CodeGen.cpp:
```diff
@@ -3569,8 +3569,13 @@ struct StoreOpConversion : public fir::FIROpConversion<fir::StoreOp> {
       if (isVolatile)
         storeOp.setVolatile_(true);
+
+      if (store.getNontemporal())
+        storeOp.setNontemporal(true);
```
Please add a lit test for this

tblah reviewed (commented):
Thanks for the updates. This is nearly there :tada:

kaviya2510 committed:
Addressed review comments: Modified getBaseOperand()function and added a testcase 'convert-nontemporal-to-llvm.fir'

tblah reviewed (approved):
Thanks for all of the updates. This now LGTM.

Small nit: a shared library build didn't work on my machine. This change fixed it for me
```diff
diff --git a/flang/lib/Optimizer/OpenMP/CMakeLists.txt b/flang/lib/Optimizer/OpenMP/CMakeLists.txt
index 22f78125c380..e31543328a9f 100644
--- a/flang/lib/Optimizer/OpenMP/CMakeLists.txt
+++ b/flang/lib/Optimizer/OpenMP/CMakeLists.txt
@@ -18,7 +18,7 @@ add_flang_library(FlangOpenMPTransforms
   LINK_LIBS
   FIRAnalysis
   FIRBuilder
-  FIRCodeGen
+  FIRCodeGenDialect
   FIRDialect
   FIRDialectSupport
   FIRSupport
```

Please wait a few days before merging in case anyone else wants to take a look.

kiranchandramohan commented:
Thanks for all the progress here.

This likely doesn’t need to be addressed in this PR, but keep in mind that some operations—such as atomic operations—may eventually lower to load and store instructions and could be missed.

```
  integer :: i
  integer :: counter
  integer, dimension(100) :: arr

  counter = 0

  !$omp simd nontemporal(arr)
  do i = 1, 100
    !$omp atomic capture
      arr(i) = counter
      counter = counter + 1
    !$omp end atomic
  end do
  !$omp end simd
end
```

kiranchandramohan commented:
> Kiran I think situations like that would be extremely difficult to address with the current design.

OK. Was just point out. No need to fix in this PR. We might want to be able to set these attributes on OpenMP operations that lower to loads or stores and handle them in translation.

kaviya2510 commented:
> Thanks for all of the updates. This now LGTM.
> 
> Small nit: a shared library build didn't work on my machine. This change fixed it for me
> 
> ```diff
> diff --git a/flang/lib/Optimizer/OpenMP/CMakeLists.txt b/flang/lib/Optimizer/OpenMP/CMakeLists.txt
> index 22f78125c380..e31543328a9f 100644
> --- a/flang/lib/Optimizer/OpenMP/CMakeLists.txt
> +++ b/flang/lib/Optimizer/OpenMP/CMakeLists.txt
> @@ -18,7 +18,7 @@ add_flang_library(FlangOpenMPTransforms
>    LINK_LIBS
>    FIRAnalysis
>    FIRBuilder
> -  FIRCodeGen
> +  FIRCodeGenDialect
>    FIRDialect
>    FIRDialectSupport
>    FIRSupport
> ```
> 
> Please wait a few days before merging in case anyone else wants to take a look.

Sure, Thanks for the approval.

kaviya2510 committed:
Merge branch 'main' into nt_to_llvmir

kaviya2510 commented:
> Thanks for all the progress here.
> 
> This likely doesn’t need to be addressed in this PR, but keep in mind that some operations—such as atomic operations—may eventually lower to load and store instructions and could be missed.
> 
> ```
>   integer :: i
>   integer :: counter
>   integer, dimension(100) :: arr
> 
>   counter = 0
> 
>   !$omp simd nontemporal(arr)
>   do i = 1, 100
>     !$omp atomic capture
>       arr(i) = counter
>       counter = counter + 1
>     !$omp end atomic
>   end do
>   !$omp end simd
> end
> ```

Thanks for sharing this testcase. I have created a ticket for the above testcase: https://github.com/llvm/llvm-project/issues/137121

kaviya2510 committed:
Added a TODO for hanlding nontemporal inside atomic construct and addressed nit comments

kaviya2510 commented:
Hi @kiranchandramohan , I have added a TODO for handling the nontemporal clause in atomic construct. Could you please review it and provide your approval?

kiranchandramohan commented on flang/lib/Optimizer/OpenMP/LowerNontemporal.cpp:
```diff
@@ -0,0 +1,82 @@
+      else if (auto storeOp = llvm::dyn_cast<fir::StoreOp>(op))
+        operand = storeOp.getMemref();
+
+      if (operand && !(fir::isAllocatableType(operand.getType()) ||
+                       fir::isPointerType((operand.getType())))) {
```
If the Allocatable and Pointer checks are for skipping the load/store of the boxes then a comment will be useful.

kaviya2510 replied:
Thanks for the comments.
Sure, I will add a comment for the same.

kiranchandramohan reviewed (approved):
Thanks @kaviya2510. Nice work.

Two comments/questions. Feel free to go ahead without further response from me.

kaviya2510 committed:
Merge branch 'main' into nt_to_llvmir

kaviya2510 committed:
Added a comment in LowerNontemporal.cpp

NimishMishra reviewed (approved):
Thanks for all the work on this. LGTM

kaviya2510 commented:
Thanks for the review @NimishMishra
